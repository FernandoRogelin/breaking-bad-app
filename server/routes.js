const routes = require('next-routes');

module.exports = routes()
  .add('login')
  .add('episodes')
  .add('characters')
  .add('quotes');
