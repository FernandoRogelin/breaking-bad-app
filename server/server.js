const Koa = require('koa');
const Router = require('koa-router');
const helmet = require('koa-helmet');
const logger = require('koa-logger');
const bodyParcer = require('koa-bodyparser');
const nextjs = require('next');

const nextRoutes = require('./routes');
const port = process.env.PORT || 8080;
const app = nextjs({ dev: process.env.NODE_ENV !== 'production' });
const nextHandler = nextRoutes.getRequestHandler(app);

app
  .prepare()
  .then(() => {
    const server = new Koa();
    const router = new Router();

    server
      .use(logger())
      .use(bodyParcer())
      .use(helmet());

    router.get('/', ctx => {
      ctx.redirect('/login');
    });

    // Handle NextJS pages
    router.get('*', async ctx => {
      await nextHandler(ctx.req, ctx.res);

      ctx.status = 200;
      ctx.respond = false;
    });

    server.use(router.routes(), router.allowedMethods());

    server.listen(port, err => {
      if (err) console.log(err);
      console.log(`Server running in port http://localhost:${port}`);
    });
  })
  .catch(ex => {
    console.log(ex.stack);
    process.exit(1);
  });
