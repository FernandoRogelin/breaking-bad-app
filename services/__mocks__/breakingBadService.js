export const getCharacters = jest.fn().mockResolvedValue({
  status: 200,
  data: [
    {
      appearance: [1, 2, 3, 4, 5],
      birthday: '09-07-1958',
      char_id: 1,
      img:
        'https://images.amcnetworks.com/amc.com/wp-content/uploads/2015/04/cast_bb_700x1000_walter-white-lg.jpg',
      name: 'Walter White',
      nickname: 'Heisenberg',
      occupation: ['High School Chemistry Teacher', 'Meth King Pin'],
      portrayed: 'Bryan Cranston',
      status: 'Presumed dead',
    },
    {
      appearance: [1, 2, 3, 4, 5],
      birthday: '09-24-1984',
      char_id: 2,
      img:
        'https://upload.wikimedia.org/wikipedia/en/thumb/f/f2/Jesse_Pinkman2.jpg/220px-Jesse_Pinkman2.jpg',
      name: 'Jesse Pinkman',
      nickname: "Cap n' Cook",
      occupation: ['Meth Dealer'],
      portrayed: 'Aaron Paul',
      status: 'Alive',
    },
    {
      appearance: [1, 2, 3, 4, 5],
      birthday: '08-11-1970',
      char_id: 3,
      img:
        'https://s-i.huffpost.com/gen/1317262/images/o-ANNA-GUNN-facebook.jpg',
      name: 'Skyler White',
      nickname: 'Sky',
      occupation: [
        'House wife',
        'Book Keeper',
        'Car Wash Manager',
        'Taxi Dispatcher',
      ],
      portrayed: 'Anna Gunn',
      status: 'Alive',
    },
    {
      appearance: [1, 2, 3, 4, 5],
      birthday: '07-08-1993',
      char_id: 4,
      img:
        'https://media1.popsugar-assets.com/files/thumbor/WeLUSvbAMS_GL4iELYAUzu7Bpv0/fit-in/1024x1024/filters:format_auto-!!-:strip_icc-!!-/2018/01/12/910/n/1922283/fb758e62b5daf3c9_TCDBRBA_EC011/i/RJ-Mitte-Walter-White-Jr.jpg',
      name: 'Walter White Jr.',
      nickname: 'Flynn',
      occupation: ['Teenager'],
      portrayed: 'RJ Mitte',
      status: 'Alive',
    },
    {
      appearance: [1, 2, 3, 4, 5],
      birthday: 'Unknown',
      char_id: 5,
      img:
        'https://upload.wikimedia.org/wikipedia/en/thumb/c/c1/Hank_Schrader2.jpg/220px-Hank_Schrader2.jpg',
      name: 'Henry Schrader',
      nickname: 'Hank',
      occupation: ['DEA Agent'],
      portrayed: 'Dean Norris',
      status: 'Deceased',
    },
  ],
});

export const getCharacterDetails = jest.fn().mockResolvedValue({
  status: 200,
  data: {
    appearance: '1,2,3',
    birthday: null,
    char_id: 31,
    img:
      'https://vignette.wikia.nocookie.net/breakingbad/images/c/c9/AdamPinkman.png/revision/latest?cb=20141117180836&path-prefix=es',
    name: 'Adam Pinkman',
    nickname: "Jesse's dad",
    occupation: 'unknown',
    portrayed: 'Michael Bofshever',
    status: 'Alive',
  },
});

export const getEpisodesList = jest.fn().mockResolvedValue({
  status: 200,
  data: [
    {
      air_date: '01-20-2008',
      characters: [
        'Walter White',
        'Jesse Pinkman',
        'Skyler White',
        'Hank Schrader',
        'Marie Schrader',
        'White White Jr.',
      ],
      episode: '1',
      episode_id: 1,
      season: '1',
      title: 'Pilot',
    },
    {
      air_date: '01-27-2008',
      characters: [
        'Walter White',
        'Jesse Pinkman',
        'Skyler White',
        'White White Jr.',
        'Krazy-8',
      ],
      episode: '2',
      episode_id: 2,
      season: '1',
      title: "Cat's in the Bag...",
    },
    {
      air_date: '02-10-2008',
      characters: [
        'Walter White',
        'Jesse Pinkman',
        'Skyler White',
        'Hank Schrader',
        'Marie Schrader',
        'White White Jr.',
      ],
      episode: '3',
      episode_id: 3,
      season: '1',
      title: "...And the Bag's in the River",
    },
    {
      air_date: '02-17-2008',
      characters: [
        'Walter White',
        'Jesse Pinkman',
        'Skyler White',
        'Hank Schrader',
        'Marie Schrader',
        'White White Jr.',
      ],
      episode: '4',
      episode_id: 4,
      season: '1',
      title: 'Cancer Man',
    },
  ],
});

export const getEpisodeDetails = jest.fn().mockResolvedValue({
  status: 200,
  data: {
    air_date: '05-03-2009',
    characters:
      'Walter White, Jesse Pinkman, Skyler White, Hank Schrader, Marie Schrader, White White Jr., Saul Goodman, Jane Margolis',
    episode: '9',
    episode_id: 16,
    season: '2',
    title: '4 Days Out',
  },
});

export const getQuotes = jest.fn().mockResolvedValue({
  status: 200,
  data: {},
});
