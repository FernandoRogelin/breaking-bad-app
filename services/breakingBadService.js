import axios from 'axios';

export const urlBase = 'https://www.breakingbadapi.com/api/';

export const getCharacters = () => axios.get(`${urlBase}characters`);

export const getCharacterDetails = id =>
  axios.get(`${urlBase}characters/${id}`);

export const getEpisodesList = () => axios.get(`${urlBase}episodes`);

export const getEpisodeDetails = id => axios.get(`${urlBase}episodes/${id}`);

export const getQuotes = () => axios.get(`${urlBase}quotes`);
