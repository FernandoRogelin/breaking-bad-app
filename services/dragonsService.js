import axios from 'axios';

export const urlBase =
  'http://5c4b2a47aa8ee500142b4887.mockapi.io/api/v1/dragon';

export const getDragons = () => axios.get(urlBase);

export const getDetailsDragon = id => axios.get(`${urlBase}/${id}`);

export const createDragon = ({ name, type, createdAt }) =>
  axios.post(urlBase, {
    name,
    type,
    createdAt,
  });

export const deleteDragon = id => axios.delete(`${urlBase}/${id}`);

export const editDragon = (id, { name, type, createdAt }) =>
  axios.put(`${urlBase}/${id}`, {
    name,
    type,
    createdAt,
  });
