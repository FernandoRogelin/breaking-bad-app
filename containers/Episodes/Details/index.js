import React, { useState, useEffect } from 'react';

import { format } from 'date-fns';
import { getEpisodeDetails } from '#/services/breakingBadService';

import {
  Wrapper,
  Title,
  Content,
  Text,
  Info,
  Header,
  Button,
  Items,
} from './styles';

export default function Details({ id, setStateDetailsPage }) {
  const [episodeDetails, setEpisodeDetails] = useState();
  const [fetched, setFetched] = useState(false);

  async function handleCharacterDetails() {
    try {
      const { data } = await getEpisodeDetails(id);

      setFetched(true);
      setEpisodeDetails(data);
    } catch (e) {
      console.log(e);
    }
  }

  useEffect(() => {
    if (!fetched) {
      handleCharacterDetails();
    }
  }, [fetched]);

  return (
    <Wrapper>
      <Header>
        <Button
          data-testid="return-episodes-page"
          onClick={setStateDetailsPage}
        >
          Return
        </Button>
        <Title>Episodes Details</Title>
      </Header>
      {episodeDetails && (
        <Content>
          {episodeDetails[0].title && (
            <Items>
              <Text>Title:</Text>
              <Info>{episodeDetails[0].title}</Info>
            </Items>
          )}
          {episodeDetails[0].season && (
            <Items>
              <Text>Season:</Text>
              <Info>{episodeDetails[0].season}</Info>
            </Items>
          )}
          {episodeDetails[0].characters && (
            <Items>
              <Text>Characters:</Text>
              <Info>{episodeDetails[0].characters}</Info>
            </Items>
          )}
          {episodeDetails[0].air_date && (
            <Items>
              <Text>Air Date:</Text>
              <Info>{format(episodeDetails[0].air_date, 'DD/MM/YYYY')}</Info>
            </Items>
          )}
        </Content>
      )}
    </Wrapper>
  );
}
