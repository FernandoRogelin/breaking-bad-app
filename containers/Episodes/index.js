import React, { useState, useEffect } from 'react';

import { Router } from '#/server/routes';
import { getEpisodesList } from '#/services/breakingBadService';
import Details from './Details';

import {
  Top,
  Name,
  Text,
  Title,
  Itens,
  Header,
  Button,
  Wrapper,
  Content,
} from './styles';

export default function Episodes() {
  const [episodes, setEpisodes] = useState([]);
  const [fetched, setFetched] = useState(false);
  const [id, setId] = useState('');
  const [isShowDetails, setIsShowDetails] = useState(false);

  async function handleEpisodesList() {
    try {
      const { data } = await getEpisodesList();

      setFetched(true);
      setEpisodes(data);
    } catch (e) {
      console.log(e);
    }
  }

  useEffect(() => {
    if (!fetched) {
      handleEpisodesList();
    }
  }, [fetched]);

  function handleDetailsPage(id) {
    setId(id);
    setIsShowDetails(!isShowDetails);
  }

  return (
    <div>
      {isShowDetails ? (
        <Details id={id} setStateDetailsPage={handleDetailsPage} />
      ) : (
        <Wrapper>
          <Header>
            <Button
              top
              data-testid="handle-characters-page"
              onClick={() => Router.pushRoute('characters')}
            >
              Return
            </Button>
            <Title>Episodes List</Title>
          </Header>
          {episodes && (
            <Content data-testid="episodes-list">
              <Top>
                <Text>Title</Text>
                <Text>Actions</Text>
              </Top>
              {episodes
                .sort((a, b) =>
                  a.title < b.title ? -1 : a.title > b.title ? 1 : 0
                )
                .map(({ title, episode_id }) => (
                  <Itens key={episode_id}>
                    <Name>{title}</Name>
                    <Button
                      onClick={() => handleDetailsPage(episode_id)}
                      data-testid="episode-details-button"
                    >
                      Episodes details
                    </Button>
                  </Itens>
                ))}
            </Content>
          )}
        </Wrapper>
      )}
    </div>
  );
}
