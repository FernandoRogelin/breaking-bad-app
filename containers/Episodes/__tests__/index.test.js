import React from 'react';
import { render, wait, fireEvent } from 'react-testing-library';
import { Router } from '#/server/routes';

Router.pushRoute = jest.fn();

import { getEpisodesList } from '#/services/breakingBadService';
import Episodes from '..';

jest.mock('#/services/breakingBadService');

describe('Episodes container', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('Episodes snapshot', () => {
    const { container } = render(<Episodes />);

    expect(container).toMatchSnapshot();
  });

  test('Inicial flow', async () => {
    const { getByTestId } = render(<Episodes />);

    await wait(() => expect(getEpisodesList).toHaveBeenCalledTimes(1));

    expect(getByTestId('episodes-list')).toBeVisible();
  });

  test('Handle character page', () => {
    const { getByTestId } = render(<Episodes />);

    fireEvent.click(getByTestId('handle-characters-page'));
  });

  test('Handle episodes details', async () => {
    const { getByTestId } = render(<Episodes />);

    await wait(() => expect(getEpisodesList).toHaveBeenCalledTimes(1));

    expect(getByTestId('episode-details-button')).toBeVisible();

    fireEvent.click(getByTestId('episode-details-button'));
  });
});
