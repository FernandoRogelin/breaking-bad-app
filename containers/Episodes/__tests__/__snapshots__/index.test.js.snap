// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Episodes container Episodes snapshot 1`] = `
<div>
  <div>
    <div
      class="styles__Wrapper-rfloh5-0 exTLaU"
    >
      <header
        class="styles__Header-rfloh5-8 fxHdmz"
      >
        <button
          class="Button-n3cyj4-0 styles__Button-rfloh5-9 dRIKXK"
          data-testid="handle-characters-page"
        >
          Return
        </button>
        <h1
          class="styles__Title-rfloh5-1 gbHtjp"
        >
          Episodes List
        </h1>
      </header>
      <div
        class="Content-vb879o-0 styles__Content-rfloh5-5 cbhcUF"
        data-testid="episodes-list"
      >
        <div
          class="styles__Top-rfloh5-2 jTRakP"
        >
          <p
            class="styles__Text-rfloh5-6 ewyQwF"
          >
            Title
          </p>
          <p
            class="styles__Text-rfloh5-6 ewyQwF"
          >
            Actions
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
